xds/internal/balancer/wrrlocality/balancer.go

```diff
@@ -0,0 +1,59 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package wrrlocality provides an implementation of the wrr locality LB policy,
+// as defined in
+// https://github.com/grpc/proposal/blob/master/A52-xds-custom-lb-policies.md.
+package wrrlocality
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// Name is the name of wrr_locality balancer.
+const Name = "xds_wrr_locality_experimental"
+
+func init() {
+	balancer.Register(bb{})
+}
+
+type bb struct{}
+
+func (bb) Name() string {
+	return Name
+}
+
+func (bb) Build(cc balancer.ClientConn, bOpts balancer.BuildOptions) balancer.Balancer {
+	return nil
+}
+
+func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	var lbCfg *LBConfig
+	if err := json.Unmarshal(s, &lbCfg); err != nil {
+		return nil, fmt.Errorf("xds: unable to unmarshal LBConfig for wrrlocality: %s, error: %v", string(s), err)
+	}
+	if lbCfg == nil || lbCfg.ChildPolicy == nil {
+		return nil, errors.New("xds: unable to unmarshal LBConfig for wrrlocality: child policy field must be set")
```

Comment:
Please look at how these error strings show up in reality. The [unmarshal code for balancer config](https://github.com/grpc/grpc-go/blob/master/internal/serviceconfig/serviceconfig.go#L74) already adds some contextual data, and we might be adding redundant data here.

Reply:
The Child Policy field is of this balancer config type you linked to, so the second error string is correct, since it's a layer above that unmarshaling. The first error string wraps information about the error returned from that Unmarshal call, if called (i.e. can also error at this configs layer) with information about this Balancer Config's layer (i.e. the Wrr Locality balancers configuration itself). I don't see redundant information being added. Thoughts?

Reply:
Hmmmm ok. Due to our tight coupling of Unmarshaling and verification, I guess I figured semantically, "failed to unmarshal" encapsulated both unmarshaling and verification due to the nature of the UnmarshalJSON() method on our internal Balancer Config. Will switch.